src/openms/source/METADATA/ID/IdentificationData.cpp

```diff
@@ -251,154 +290,190 @@ namespace OpenMS
+        throw Exception::IllegalArgument(__FILE__, __LINE__,
+                                         OPENMS_PRETTY_FUNCTION, msg);
+      }
     }
 
-    return insertIntoMultiIndex_(parent_molecules_, parent,
-                                 parent_molecule_lookup_);
+    return insertIntoMultiIndex_(parents_, parent,
+                                 parent_lookup_);
   }
 
 
-  void IdentificationData::registerParentMoleculeGrouping(
-    const ParentMoleculeGrouping& grouping)
+  void IdentificationData::registerParentGroupSet(const ParentGroupSet& groups)
   {
-    checkAppliedProcessingSteps_(grouping.steps_and_scores);
-
-    for (const auto& group : grouping.groups)
+    if (!no_checks_)
     {
-      checkScoreTypes_(group.scores);
+      checkAppliedProcessingSteps_(groups.steps_and_scores);
 
-      for (const auto& ref : group.parent_molecule_refs)
+      for (const auto& group : groups.groups)
       {
-        if (!isValidHashedReference_(ref, parent_molecule_lookup_))
+        checkScoreTypes_(group.scores);
+
+        for (const auto& ref : group.parent_refs)
         {
-          String msg = "invalid reference to a parent molecule - register that first";
-          throw Exception::IllegalArgument(__FILE__, __LINE__,
-                                           OPENMS_PRETTY_FUNCTION, msg);
+          if (!isValidHashedReference_(ref, parent_lookup_))
+          {
+            String msg = "invalid reference to a parent sequence - register that first";
+            throw Exception::IllegalArgument(__FILE__, __LINE__,
+                                             OPENMS_PRETTY_FUNCTION, msg);
+          }
         }
       }
     }
 
-    parent_molecule_groupings_.push_back(grouping);
+    parent_groups_.push_back(groups);
 
     // add the current processing step?
     if ((current_step_ref_ != processing_steps_.end()) &&
-        (grouping.steps_and_scores.get<1>().find(current_step_ref_) ==
-         grouping.steps_and_scores.get<1>().end()))
+        (groups.steps_and_scores.get<1>().find(current_step_ref_) ==
+         groups.steps_and_scores.get<1>().end()))
     {
-      parent_molecule_groupings_.back().steps_and_scores.push_back(
+      parent_groups_.back().steps_and_scores.push_back(
         IdentificationDataInternal::AppliedProcessingStep(current_step_ref_));
     }
   }
 
 
-  IdentificationData::QueryMatchRef
-  IdentificationData::registerMoleculeQueryMatch(const MoleculeQueryMatch&
-                                                 match)
+  IdentificationData::AdductRef
+  IdentificationData::registerAdduct(const AdductInfo& adduct)
   {
-    if (const IdentifiedPeptideRef* ref_ptr =
-        boost::get<IdentifiedPeptideRef>(&match.identified_molecule_ref))
+    // @TODO: require non-empty name? (auto-generate from formula?)
+    auto result = adducts_.insert(adduct);
+    if (!result.second && (result.first->getName() != adduct.getName()))
```

Comment:
when does this case happen? e.g., what needs to equal? e.g. having two different adducts with same formula should be ok I guess

Reply:
Currently different adducts with the same sum formula and charge aren't allowed (i.e. they're considered the same adduct). This could be changed, but would introduce ambiguities because in practice adducts are matched by mass and charge. So if you get a match for one, does that also imply a match for the other (with same formula/charge)?

See `AdductCompare` in ObservationMatch.h (https://github.com/hendrikweisser/OpenMS/blob/98656f05c5be92a6be822e65f6c6f450f37e4a83/src/openms/include/OpenMS/METADATA/ID/ObservationMatch.h#L56-L66).

A possible alternative would be to keep a list of names for the "same" adduct.

Reply:
Hmm … I could imagine that many structures that could be potential affects have the same formula

Reply:
@oliveralka can you think about the implications?